polytracker/mapping.py

```diff
@@ -3,156 +3,191 @@
-        for source, tree in sources.items():
-            unused_bytes = IntervalTree([Interval(0, source.size)])
-            for interval in tree:
-                unused_bytes.chop(interval.begin, interval.end)
-            for interval in sorted(unused_bytes):
-                yield TaintedRegion(
-                    source=source,
-                    offset=interval.begin,
-                    length=interval.end - interval.begin,
-                )
+    def __init__(self, f: TDFile):
+        self.tdfile: TDFile = f
+
+    def mapping(self) -> Dict[OffsetType, Set[OffsetType]]:
+        raise NotImplementedError()
+
+    def marker_to_ranges(self, m: bytearray) -> List[CavityType]:
+        ranges = []
+        start = None
+        for i, v in enumerate(m):
+            if v == 0:
+                if start is None:
+                    start = i
+            if v == 1:
+                if start is not None:
+                    ranges.append((start, i))
+                    start = None
+        if start is not None:
+            ranges.append((start, len(m) - 1))
+        return ranges
+
+    def file_cavities(self) -> Dict[Path, List[CavityType]]:
+        seen: Set[int] = set()
+
+        def source_labels_not_affecting_cf(label: int) -> Generator[int, None, None]:
+            stack = [label]
+            while len(stack) > 0:
+                lbl = stack.pop()
+
+                if lbl in seen:
+                    continue
+
+                seen.add(lbl)
+
+                n = self.tdfile.decode_node(lbl)
+
+                if n.affects_control_flow:
+                    continue
+
+                if isinstance(n, TDSourceNode):
+                    yield lbl
+
+                elif isinstance(n, TDUnionNode):
+                    nl = self.tdfile.decode_node(n.left)
+                    if not nl.affects_control_flow:
+                        if isinstance(nl, TDSourceNode):
+                            yield n.left
+                        else:
+                            stack.append(n.left)
+
+                    nr = self.tdfile.decode_node(n.right)
+                    if not nr.affects_control_flow:
+                        if isinstance(nr, TDSourceNode):
+                            yield n.right
+                        else:
+                            stack.append(n.right)
+
+                elif isinstance(n, TDRangeNode):
+                    for rl in range(n.first, n.last + 1):
+                        # NOTE: One could skip decoding here, but then we could end up with really long ranges
+                        # being added the labels that really does nothing except cause overhead...
+                        rn = self.tdfile.decode_node(rl)
+                        if rn.affects_control_flow:
+                            continue
+                        if isinstance(rn, TDSourceNode):
+                            yield rl
+                        else:
+                            stack.append(rl)
```

Comment:
Maybe we could add the `(rl, rn)` here and have a check on line 54 that checks `isinstance(lbl, int)` before doing `decode_node`? That would prevent decoding a taint node twice. Might be worth experimenting if it increases performance.

Reply:
Also, if this is done, one should probably do the same for `n.left` and `n.right` when it is a TDUnionNode.

Reply:
I'm not sure I understand what you mean here. Can you write a suggestion with the changes?

Reply:
I've made a few suggestions. Not sure it is valid python code, but conceptually that's what I mean. Makes sense?